File: 2023/erlang/day03a/day03a.erl
-module (day03a).
-export ([main/1]).

usage() ->
    Module = "day03a.erl",
    io:format("usage: escript ~p <file>~n", [Module]),
    halt(1).

build_numbers(Contents) ->
    element(2, lists:foldl(fun(Line, {Row, NumberLocs}) ->
        {Row + 1, element(5, lists:foldl(fun(Ch, {Col, ScanningNumber, Number, CurrentPos, InnerNumberLocs}) ->
                case ScanningNumber of
                    true ->
                        if Ch >= $0 andalso Ch =< $9 ->
                            if Col + 1 == length(Line) ->
                                {Col + 1, false, "", {-1, -1}, maps:put(CurrentPos, Number ++ [Ch], InnerNumberLocs)}
                            ; true ->
                                {Col + 1, true, Number ++ [Ch], CurrentPos, InnerNumberLocs}
                            end
                        ; true ->
                            {Col + 1, false, "", {-1, -1}, maps:put(CurrentPos, Number, InnerNumberLocs)}
                        end;
                    false ->
                        if Ch >= $0 andalso Ch =< $9 ->
                            {Col + 1, true, [Ch], {Row, Col}, InnerNumberLocs}
                        ; true ->
                            {Col + 1, false, "", CurrentPos, InnerNumberLocs}
                        end
                end
            end, {0, false, "", {-1, -1}, NumberLocs}, Line))}
    end, {0, #{}}, string:split(string:trim(Contents), "\n", all))).

build_parts(Contents) ->
    element(2, lists:foldl(fun(Line, {Row, PartLocs}) ->
        {Row + 1, element(2, lists:foldl(fun(Ch, {Col, InnerPartLocs}) ->
            if not (Ch >= $0 andalso Ch =< $9) andalso Ch /= $. ->
                {Col + 1, maps:put({Row, Col}, Ch, InnerPartLocs)}
            ; true ->
                {Col + 1, InnerPartLocs}
            end
        end, {0, PartLocs}, Line))}
    end, {0, #{}}, string:split(string:trim(Contents), "\n", all))).

check_parts(NumberLocs, PartLocs) ->
    lists:foldl(fun (NumberLoc, Result) ->
            NumberRow = element(1, NumberLoc),
            NumberColFirst = element(2, NumberLoc),
            NumberColLast = element(2, NumberLoc) + string:length(maps:get(NumberLoc, NumberLocs)) - 1,
            FoundAdjacent = lists:foldl(fun (NumberCol, Found) ->
                    if Found ->
                        true
                    ; true ->
                        lists:foldl(fun (Neighbor, FoundAdj) ->
                                AdjacentPos = { NumberRow + element(1, Neighbor), NumberCol + element(2, Neighbor)},
                                FoundAdj or maps:is_key(AdjacentPos, PartLocs)
                            end, false, [{-1, -1}, {-1, 0}, {-1, 1}, {0, -1}, {0, 1}, {1, -1}, {1, 0}, {1, 1}])
                    end
                end, false, lists:seq(NumberColFirst, NumberColLast)),
            case FoundAdjacent of
                true -> Result + element(1, (string:to_integer(maps:get(NumberLoc, NumberLocs))));
                false -> Result
            end
        end, 0, maps:keys(NumberLocs)).

process_input(Contents) ->
    NumberLocs = build_numbers(Contents),
    PartLocs = build_parts(Contents),
    check_parts(NumberLocs, PartLocs).

main([Filename]) ->
    {ok, File} = file:read_file(Filename),
    Contents = unicode:characters_to_list(File),
    Result = process_input(Contents),
    io:format("result = ~w~n", [Result]);
main(_) ->
    usage().

File: 2023/go/day03a/day03a.go
package main

import (
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
	"unicode"
)

type Position struct {
	Row int
	Col int
}

func usage(progname string) {
	log.Fatalf("usage: %s <file>\n", progname)
	os.Exit(1)
}

func buildNumbers(contents string) map[Position]string {
	numberLocs := make(map[Position]string)
	lines := strings.Split(contents, "\n")
	scanningNumber := false
	currentPos := Position{Row: -1, Col: -1}
	number := []rune{}
	for row, line := range lines {
		for col, ch := range line {
			if scanningNumber {
				if unicode.IsDigit(ch) {
					number = append(number, ch)
				} else {
					numberLocs[currentPos] = string(number)
					scanningNumber = false
					number = []rune{}
				}
			} else {
				if unicode.IsDigit(ch) {
					number = append(number, ch)
					currentPos = Position{Row: row, Col: col}
					scanningNumber = true
				}
			}
		}
		if scanningNumber {
			numberLocs[currentPos] = string(number)
			scanningNumber = false
			number = []rune{}
		}
	}
	return numberLocs
}

func buildParts(contents string) map[Position]rune {
	partLocs := make(map[Position]rune)
	lines := strings.Split(contents, "\n")
	for row, line := range lines {
		for col, ch := range line {
			if !unicode.IsDigit(ch) && ch != '.' {
				partLocs[Position{Row: row, Col: col}] = ch
			}
		}
	}
	return partLocs
}

func checkParts(numberLocs map[Position]string, partLocs map[Position]rune) uint32 {
	result := uint32(0)
	for numberLoc, numberStr := range numberLocs {
		adjacentCount := 0
		numberRow := numberLoc.Row
		numberColStart := numberLoc.Col
		numberColLast := numberLoc.Col + len(numberStr)
		for numberCol := numberColStart; numberCol < numberColLast; numberCol++ {
			neighbors := [...]Position{
				{Row: -1, Col: -1}, {Row: -1, Col: 0}, {Row: -1, Col: 1},
				{Row: 0, Col: -1}, {Row: 0, Col: 1},
				{Row: 1, Col: -1}, {Row: 1, Col: 0}, {Row: 1, Col: 1},
			}
			for partLoc, _ := range partLocs {
				for _, neighbor := range neighbors {
					pos := Position{Row: numberRow + neighbor.Row, Col: numberCol + neighbor.Col}
					if partLoc.Row == pos.Row && partLoc.Col == pos.Col {
						adjacentCount++
						break
					}
				}
			}
		}
		if adjacentCount != 0 {
			number, _ := strconv.Atoi(numberStr)
			result += uint32(number)
		}
	}
	return result
}

func process(contents string) uint32 {
	numberLocs := buildNumbers(contents)
	partLocs := buildParts(contents)
	return checkParts(numberLocs, partLocs)
}

func main() {
	if len(os.Args) < 2 {
		usage(os.Args[0])
	}
	filename := os.Args[1]
	buf, err := os.ReadFile(filename)
	if err != nil {
		log.Fatal(err)
	}
	contents := string(buf[:])
	result := process(contents)
	fmt.Printf("result = %d\n", result)
}
